### PR TITLE
 SAML SLO support

### PR DESCRIPTION
**Pull request by andreas-kupries**
## Issue: 

See #38494

This work is co-dependent on the UI work tracked at https://github.com/rancher/dashboard/issues/10941
 
## Problem

SURE 3572
 
## Solution

1. Extended `AuthConfig`, `SamlConfig` with the proposed flags about SLO (`supported`, `enabled`, `forced`).
    1. Based on the CRD setup the `supported` flag might be nonsense.
    2. As in, cannot be set into the initial AuthConfig CR instances. UI may have to simply know that only the SAML providers support SLO, and none of the others.
2. New structures `SamlConfigLogoutInput`, and `...Output`. Same fields as the known `SamlConfigTest...` structures. Hold the request/response data from/to the UI for the `logoutAll` action (see below).
3. The `tokens` API should export a new action `logoutAll`.
4. Basic implemention of the logout flow. Compiles, untested.
5. Linkage between token manager and saml to invoke the flow from the frontend

KNOWN ISSUES: Does not guard against call of regular logout when SLO is forced.
Does guard against forced but not enabled, and call to logout-all when not enabled.

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

**Comment by andreas-kupries** [comment]
Status report after a few days of working on implementing the thing:

The main visualization of the necessary workflow is taken from
http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02_html_m50a2ba3e.gif
The same is described at https://xacmlinfo.org/2013/06/28/how-saml2-single-logout-works/

The main take I get from both is that the SP who initiated the SLO is __not__ notified by the IdP that the user was logged out.

Okta seems to follow that principle, based on what I experienced with it today. I.e. I see a plain POST to Rancher's `.../saml/slo` endpoint, and based on the fact that I can see it in the browser's debug console I believe that this was indeed done by redirection from the initial norman action (`logoutAll`). Unfortunately this redirect looks to have lost tall he Cookies used by the SAML code to store the flow state. I am unsure if that may be because the response contained in the redirect is fail, not a success. Although I cannot see how the browser would know that. Only thing I can think of is that Okta unsets these, maybe all, cookies. The failure is not due to a bogus user reference. Looking at the Okta Events the logout request is recorded as such, and its recognizes the user. It claims to have found an `Invalid signature`.

That is weird because the signature on the logout request should be generated through the same code as for SSO, and use the same cert/key, and SSO is, well, successful. Given my `should` there is doubt on that, mainly because different key and such is the only way I can see this failure to happen. The code is still the same, service provider name is same, so same SP structures, same key data. 

Going back to KC, which I struggled with before trialing Okta today it seems that this IdP wants to notify the initiating SP also, possibly directly from itself. The initial failures I saw with it said `Unable to finish logout because XXX endpoint is not set`. With an endpoint set for it the other issues I seem to have look to be me not properly processing/responding to that notification.

It should be noted that  the XXX endpoint is generally needed, namely to handle IdP initiated logouts, i.e. when Rancher and other apps share the IdP and users, and a user in the other apps performs a SLO.

Thus, if I can implement this endpoint properly then it should be possible to use KC completely.

This is the state of the work as of commit 582687e81dff9b1e5b8bb6f6de68e8ad0ab1431b.

Thanks to @aalves08 for providing access to a remote Okta IdP to work with.

**Comment by richard-cox** [comment]
@andreas-kupries At some point next week could you provide an linux/amd64 image of the changes in this PR? I'll be taking over from Alex whilst he's away and noticed he was using arm64

**Comment by andreas-kupries** [comment]
Managed a full logout with KC now.

That said, after that the UI does a number of things I was unable to track, and then landed on
> https://tagetarl:8005/auth/login?err=An%20error%20occurred%20logging%20in.%20Please%20try%20again.

in the end. In other words, there may still be something more to do UI side.

Side note: I have to keep the haering SAML package. The logout response is deflate-compressed. haering handles that ok. crewjam does not try to decompress, tries to read the compressed string as XML and then fails with a bad utf-8 error.

**Comment by andreas-kupries** [comment]
Status report: SLO works for Keycloak and OKTA now.

  - The keyword is `detached signature`
    - Not handled by crewjam/saml on receiving responses :arrow_right: KC issue
    - Not generated by crewjam/saml for logout redirect requests :arrow_right: OKTA issue
  - Fixed now
    - 99%: https://github.com/crewjam/saml/pull/449
    - 1% remainder: https://github.com/crewjam/saml/pull/449#discussion_r1654477177
  - Commits to `rancher/saml`:
    - https://github.com/rancher/saml/commit/e1e76fd2b4b6fa44c3264311f093cc7bce14d045
    - https://github.com/rancher/saml/commit/a10be19085301a0f36419b77a68cab400eaf9a7c

File: pkg/auth/providers/saml/saml_client.go
Comment (crobby) [inline pkg/auth/providers/saml/saml_client.go:331]: Has this fork been done yet?  i.e. Can we remove the TODO?

**Review by crobby — COMMENTED** [review]
General comment.  There seems to be quite a number of debug logs.  Our debug logs overall in Rancher are super dense.  
The answer may be, "yes", and that is fine, but have you considered removing any of these that might produce lots of output that isn't very high-value?

File: pkg/auth/providers/saml/saml_client.go
Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_client.go:331]: yes, fork is done, todo can go.

Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_client.go]: Done with commit [7437e4747656226f7291311d9ae2fa4b9ff79e58]

**Comment by andreas-kupries** [comment]
> General comment. There seems to be quite a number of debug logs.

I tend towards higher amount of logging by default.

>  Our debug logs overall in Rancher are super dense. The answer may be, "yes",
> and that is fine, but have you considered removing any of these that might
> produce lots of output that isn't very high-value?

Weeded a bit, result is in commit [2a3bf9bcdf924cd4445bd43feef689edd5dfa5d9]

File: pkg/auth/providers/saml/saml_client.go
Comment (pjbgf) [inline pkg/auth/providers/saml/saml_client.go:123]: Maybe?
```suggestion
		return fmt.Errorf("invalid SAML configuration: cannot force SLO if not enabled")
```

File: pkg/auth/providers/saml/saml_client.go
Comment (pjbgf) [inline pkg/auth/providers/saml/saml_client.go:324]: It is probably a good idea to ensure the values in `errorMsg` are encoded correctly.

```suggestion
		rURL, err := url.Parse(redirectURL)
		...
		params := rURL.Query()
		params.Add("errorCode", "500")
		params.Add("errorMsg", err.Error())
		
		rURL.RawQuery = params.Encode()
```

then we would use `rURL.String()` instead of `redirectURL`.

File: pkg/auth/providers/saml/saml_handlers.go
Comment (pjbgf) [inline pkg/auth/providers/saml/saml_handlers.go:43]: Any information here we may want to redact from the logs?

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:372]: Agreed.

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:375]: Agreed that we should modernize this piece of old code a bit,

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:395]: Agreed.

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:385]: I had forgotten about its presence actually. ..,
The `sloForced` info is not available here actually.

Implementing this guard requires another callback, like the `onLogoutAll` to get into the SAML context and make the check.

Not sure if we want that. 

On the other hand, if SLO is pushed out to 2.10, I would have the time to implement that without stressing over getting the PR merged (into 2.9).

File: pkg/auth/providers/saml/saml_client.go
Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_client.go:324]: Looks sensible. Note that the `ServeHTTP` code path handling the ACS endpoint has similar code, should be changed analogously.

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:372]: Done with commit [4dfbc37b8d580c817c3234ad32f946db4bdebb20]

File: pkg/auth/providers/saml/saml_client.go
Comment (andreas-kupries) [inline pkg/auth/providers/saml/saml_client.go]: @pjbgf 
This needs fixup ... And having to deal with that I am now wondering how to handle an url parse error here. 
We will not have a proper `rURL`, only the `redirectURL`.
Which is likely bad.
I suspect at that point we have no choice but to go with string ops.
IIRC there might be a an encoder function for params we can use outside of URL methods ...
... https://pkg.go.dev/net/url#QueryEscape I believe.

Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_client.go]: The latest commit has the proposed change.

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go]: Ok, implemented this now.
Commit [37152f4ec6bbd5d92019b2102a2af29a10bad044].

File: go.mod
Comment (macedogm) [inline go.mod:9]: What is our long term plan regarding this fork, please? Was this validated with the Architecture team? Just asking to understand our goals and if we plan to eventually go back to the original upstream version or not.

I see in our fork that we added some fixes and open (not merged) upstream PRs. Are we working with upstream to get those merged?

My concern is that this is a very sensitive library in terms of security, and it already had 5 CVEs - https://github.com/crewjam/saml/security . By going with the fork we might lose track of future CVE fixes.

Comment (andreas-kupries) [reply go.mod:9]: > Are we working with upstream to get those merged?

The PRs for the fixes we use look to have been ignored for more than a year, both.
And the project has lots of other open PRs which are not attended to.
It feels as if the entire project has fallen out of maintenance.
As such I am unsure how we could work with upstream to get these merged.

> What is our long term plan regarding this fork, please?

Personally I would push for us to search for a better maintained replacement.

> Was this validated with the Architecture team?

Who should I talk to ?

Comment (macedogm) [reply go.mod:9]: @mattfarina WDYT about the situation described by Andreas, please?

Comment (mattfarina) [reply go.mod:9]: @andreas-kupries I have a couple questions:

1. Do we have a documented list of changes we are carrying and the reason for each of them? I ask because 1) I am looking to understand the situation and 2) I've run into problems with other forks where we had lost that information and no longer knew.
2. Do we know of any alternative saml packages that are well maintained and stable? I suspect the answer is "no" because we have not looked but I wanted to ask. Just in case.

Comment (andreas-kupries) [reply go.mod:9]: There is currently no documented list of changes, except through the commit messages in the fork.
I am ok with writing such a list. Where should it go ? (Just) in the forked repo, or elsewhere too ?

I know of https://pkg.go.dev/github.com/russellhaering/gosaml2 because during the search
for the OKTA/KC issues it was used in part. The parts were removed again when the first crewjam patch was added,  fixing the mishandling of compressed data.

Some other packages I saw were explicitly marked as experimental, incomplete, etc.

Comment (andreas-kupries) [reply go.mod:9]: Here is a proposed list of changes, to be added to the README.md of the `rancher/saml` fork of `crewjam/saml`

---
# FORK information

This fork of the https://github.com/crewjam/saml repository is rooted in commit
a32b643a25a46182499b1278293e265150056d89 of the origin repository.

This is the head of the `main` branch at the time of writing this document, containing version
`v0.4.14` of the package, i.e. commit b07b16cf83c4171d16da4d85608cb827f183cd79, plus a number of
version bumps for dependent modules, via dependabot.

Compared to the root the fork contains the following changes

  1. Application of pull request https://github.com/crewjam/saml/pull/491

     This fixes the mishandling of compressed data. The XML parser always operated on the input data
     even if it was compressed. While decompression was performed the resulting data was ignored and
     not used.

  1. Application of pull request https://github.com/crewjam/saml/pull/449

     This adds the missing verification of detached signatures to the handling of logout responses.

  1. Beyond the above the work of item 2 was modified/extended to allow an embedded signature to be
     missing when a good, i.e. verified detached signature is found.

     This means that the code now accepts the case of detached signature without embedded signature
     as valid. This is what Keycloak provides.

  1. Going further the work of item 2 was used as template to add the generation of detached
     signatures to logout requests generated by the package.

     Such a missing detached signature was the cause of OKTA not accepting our logout requests. The
     embedded signature had no meaning.

Comment (andreas-kupries) [reply go.mod:9]: And added to the fork, with commit https://github.com/rancher/saml/commit/51375381b65c352cfe0e7635aefb20c27dc5be80

Comment (macedogm) [reply go.mod:9]: So just for my own understanding, https://pkg.go.dev/github.com/russellhaering/gosaml2 doesn't contain all the code and fixes that we need without having the need to still fork it too, right?

Note that our fork https://github.com/rancher/saml doesn't have `main` as the default branch, instead if points to `v0.0.1-rancher1` which can be confusing. Would be good to update the default branch if possible, please.

Comment (andreas-kupries) [reply go.mod:9]: > https://pkg.go.dev/github.com/russellhaering/gosaml2

This package came up during work here as a possible alternative to `crewjam/saml`, which I did not investigate deeply enough to know if it provides everything we need, or not. I saw a fork of `crewjam/saml` as the less risky path than having to replace it with something completely new.

> Would be good to update the default branch if possible, please.

Checking, I find that I do not have the admin permissions necessary to change the default path. Who should I talk to get the change done ?

Comment (macedogm) [reply go.mod:9]: Thanks for the explanation, Andreas.

> I saw a fork of crewjam/saml as the less risky path than having to replace it with something completely new.

Putting this into perspective, it does make sense.

The security implications, ownership and maintainability of such forks, that and others, will eventually be addressed in the Engineering Handbook (CC @mattfarina).

> Checking, I find that I do not have the admin permissions necessary to change the default path. Who should I talk to get the change done ?

You must open a GH issue for EIO.

Comment (andreas-kupries) [reply go.mod:9]: Ok.

Comment (andreas-kupries) [reply go.mod:9]: Done: https://github.com/rancherlabs/eio/issues/2429

Comment (andreas-kupries) [reply go.mod:9]: EIO is done - https://github.com/rancherlabs/eio/issues/2429#issuecomment-2209334653

**Comment by samjustus** [comment]
for posterity: security needs to finish pentesting before we can merge

**Comment by andreas-kupries** [comment]
The security review has completed as ok and closed.
I have the go ahead to merge.

__This is deferred until 2.10 is open, i.e. 2.9 is released.__

**Comment by andreas-kupries** [comment]
Relevant flaky test: https://github.com/rancher/rancher/issues/42248

**Review by crobby — APPROVED** [review]

File: pkg/auth/tokens/manager.go
Comment (bigkevmcd) [inline pkg/auth/tokens/manager.go:385]: Instead of `fmt.Sprintf`, you could just call `err.Error()` ?

File: pkg/auth/data/authconfig_data.go
Comment (bigkevmcd) [inline pkg/auth/data/authconfig_data.go:23]: Feels like this could be better expressed as two functions?

`addAuthConfig` and `addAuthConfigWithSLO` ?

**Review by bigkevmcd — APPROVED** [review]
I'm fine with this change, and I've been able to test it against Keycloak (with Backchannel logout URL).

But there is some tidying that could be done to make it easier for the next person to work on this?

File: pkg/auth/data/authconfig_data.go
Comment (andreas-kupries) [reply pkg/auth/data/authconfig_data.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:77]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/auth/tokens/manager.go
Comment (andreas-kupries) [reply pkg/auth/tokens/manager.go:98]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/auth/providers/saml/saml_provider.go
Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_provider.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/auth/providers/saml/saml_provider.go
Comment (andreas-kupries) [reply pkg/auth/providers/saml/saml_provider.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (andreas-kupries) [reply pkg/apis/management.cattle.io/v3/authn_types.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (andreas-kupries) [reply pkg/apis/management.cattle.io/v3/authn_types.go]: done. commit 3b5168e2824999c0f569ba78924028cf7a4af3d9

**Comment by andreas-kupries** [comment]
## Root Cause

A change request to support logging a User out of the session held by the configured external auth provider (EAP), and thus out of all applications, instead of just out of Rancher itself. This last meant that when logging back into Rancher the still-open session in the EAP allowed for quick login, without having to run through full authentication again. Confusing several users which expected to fully re-authenticate. Despite the notification on regular logout that the EAP session may be retained.

## What was fixed, or what change have occurred

Several, not all,  EAP now support an `LogoutAll` option and action.
The supporting EAP are the SAML variants on offer.

__Note__ that the following changes are in the Dashboard, not in the Backend.

Supporting `LogoutAll` means that when such an EAP is configured and activated the Admin
can configure the checkboxes
- `LogoutAllEnabled` and
- `LogoutAllForced`.

Checking `LogoutAllEnabled` causes the UI to offer the user the choice between regular logout and logout all.
Additionally checking `LogoutAllForced` causes the UI to not offer regular logout anymore, only logout all.
And as the sole choice no actual choice is offered to the user. Logout is logout all.
Note that `Forced` cannot be checked if `Enabled` is not checked.

The backend sees these configuration flags as well and will react with errors should the dashboard try to
- invoke `logout all` when not enabled.
- invoke `logout` when `logout all` is forced.

## Areas or cases that should be tested

- Dashboard
  - EAP configuration
    - Allow configuration of `logout all` for SAML providers.
    - Do not allow configuration of `logout all` for any other providers.
  - Logout
    - Use plain `logout`, with no choice, when `logout all` is not enabled (SAML) or not supported (any else).
    - Offer choice (dialog, IIRC) between `logout` and `logout all` when `logout all` is enabled.
    - Use `logout all`, with no choice, when `logout all` is enabled and forced.

## What areas could experience regressions

- Dashboard logout.
- EAP configuration.

## Are the repro steps accurate/minimal?

**Pull request merged**
